### DISCUSSION TITLE
Audio Ouptut channel Map Suddenly Changed

### DISCUSSION THREAD

**Discussion by OrgelSpieler** [GrandOrgue/grandorgue#928] [Help] [answered]
I've been running GO with an RME HDSPe AIO multichannel Card with RME's ASIO drivers flawlessly for over three weeks (See #900 for historical reference on my journey to ASIO).  Here is the Audio Output Map I had been successfully using.  It shows how the GO Output Mapping is directed to RME's output channels:

**TABLE 1**
Audio Output Ch___________ RME Playback CH
01 (Great Foundations L) --> 1 (Analog)
02 (Great Foundations R) --> 2 (Analog)
03 (Choir Foundations L) --> 3 (Analog)
04 (Choir Foundations R) --> 4 (Analog)
05 (Swell Foundations L) --> 5 (Analog)
06 (Swell Foundations r) --> 6 (Analog)
07 Not Used ---------------> 7 Phones L
08 Not Used ---------------> 8 Phones R
09 Not Used ---------------> 9 AES L
10 Not Used -------------> 10 AES R
11 (Pedal L) --------------> 11 (SPDIF L)
12 (Pedal R) -------------> 12 (SPDIF R)
13 (Mixtures L) ----------> 13 (ADAT 1)
14 (Mixtures R) ----------> 14 (ADAT 2)
15 (Reeds L) -------------> 15 (ADAT 3)
16 (Reeds R) -------------> 16 (ADAT 4)
17 (2', 1' L) ---------------> 17 (ADAT 5)
18 (2', 1' R) ---------------> 18 (ADAT 6)
19 (Ant Trumpets L) -----> 19 (ADAT 7)
20 (Ant Trumpets R) -----> 20 (ADAT 8)

Today, I decided to experiment a bit with the convolution reverb (I have done this before with the ASIO driver).  As soon as I enabled the reverb, the sound became very hallow.  I disabled the reverb and the hallow sound continued without the reverb effects.  A little analysis showed that my keyboard foundation, mixture and pedal ranks (channels 1-14) were no longer playing.  Restarting GO, restarting the computer, re-installing GO (over the existing installation) and re-installing the RME driver did not alter the result.  I dug a little further and found that the ranks that were sounding were being mapped as shown:

**TABLE 2**
Audio Output Ch___________ RME Playback CH
15 (Reeds L) -------------> 01 (Analog 1)
16 (Reeds R) -------------> 02 (Analog 2)
17 (2', 1' L) ---------------> 03 (Analog 3)
18 (2', 1' R) ---------------> 04 (Analog 4)
19 (Ant Trumpets L) -----> 05 (Analog 5)
20 (Ant Trumpets R) -----> 06 (Analog 6)

I tried the defualt mapping (Stereo - 2 channels) and all of the ranks began playing again.  I went back to the ASIO driver, trying the PA and RT modes and lost the previous ranks specified.  I tried various windows drivers to the extent that I could map channels and they all worked correctly.  I went back to the ASIO driver and I continued to be missing the ranks specified.

Through all the experimentation,  I noticed that when using the ASIO driver, the output to input mapping was shifted a consistent amount.  So I got curious and set up the Audio Output as shown below:

**TABLE 3**
Audio Output Ch___________ RME Playback CH
01 Not Used
02 Not Used
03 Not Used
04 Not Used
05 Not Used
06 Not Used
07 Not Used
08 Not Used
09 Not Used
10 Not Used
11 Not Used
12 Not Used
13 (Great Foundations L) --> 1 (Analog)
14 (Great Foundations R) --> 2 (Analog)
15 (Choir Foundations L) --> 3 (Analog)
16(Choir Foundations R) --> 4 (Analog)
17 (Swell Foundations L) --> 5 (Analog)
18 (Swell Foundations r) --> 6 (Analog)
19 Not Used ---------------> 7 Phones L
20 Not Used ---------------> 8 Phones R
21 Not Used ---------------> 9 AES
22 Not Used -------------> 10 AES
23 (Pedal L) --------------> 11 SPDIF
24 (Pedal R) -------------> 12 SPDIF
25 (Mixtures L) ----------> 13 (ADAT 1)
26 (Mixtures R) ----------> 14 (ADAT 2)
27 (Reeds L) -------------> 15 (ADAT 3)
28 (Reeds R) -------------> 16 (ADAT 4)
29 (2', 1' L) ---------------> 17 (ADAT 5)
30 (2', 1' R) ---------------> 18 (ADAT 6)
31 (Ant Trumpets L) -----> 19 (ADAT 7)
32 (Ant Trumpets R) -----> 20 (ADAT 8)

With this Audio Output Map all of the ranks now sound and all ranks are directed to the correct speaker channel, however it creates a long Menu Bar.  Note that this working configuration starts on Channel 13 as opposed to the partial sounding mapping that started at channel 15 (Table 2).  I cannot explain the discrepancy.

While this is working, something is obviously amiss.  I figure something was written to the registry that has caused everything to skew.  I'm sure if I did a clean installation, everything would return to normal.  Any ideas on how to fix this without starting over?

**Comment by oleg68** [comment]
@OrgelSpieler 

1. You don't have to reinstall GrandOrgue for making a clean installation. You just have to delete `<User home directory>\AppData\Roaming\GrandOrgueConfig` file. After that all your settings will be reset
2. Please remove the settings file and make the setup from scratch
3. Check that everything is all right and make a backup copy of `GrandOrgueConfig`
4. Try to reproduce your issue with enabling the convolution reverb
5. If the channel shift occure, make a bug report
6. Restore the saved `GrandOrgueConfig` file

**Reply by OrgelSpieler** [reply]
I started from scratch and ended with the same problem without enabling the reverb.  I had to start the new configuration at audio output 13 to get anything to sound.  I ended up with the same map as before.  

What does that suggest?

**Reply by oleg68** [reply]
It looks like the ASIO driver itself remaps the channels. Do you have any another program to test ASIO output channels,

**Reply by OrgelSpieler** [reply]
I have used a small ASIO test generator (http://www.tropicalcoder.com/ASIOTestSigGen.htm) to test the outputs.  It lets me select any channel I want.  It directs to the correct channel.  There is no shifting.  

Do you think that is an adequate test?

**Reply by oleg68** [reply]
> Do you think that is an adequate test?

Yes, I do.

**Reply by OrgelSpieler** [reply]
I believe this might be a Windows problem.  As I mentioned earlier, I have two RME cards installed in the system.  I use one for GO and the other for a different application.  When I tried the ASIO driver for the first time, I had the other card disabled.  After I got the GO installation running with my 16 ASIO channels, I enabled the other card and everything continued to work fine until the reverb hiccup.  All of a sudden my channels of sound that had been channels 1- 20 (4 were unused) were now mapped as 13-24.  

This evening I loaded Hauptwerk.  When it mapped the ASIO channels by default, it mapped channels 1-12 to my "other" card and 13 and ff to the Card I use for my organ.  That caused the lightbulb to go off in my head as I think the ordering has been determined by Windows.  I need to do a little more experimentation to verify this is the case.  Why the reverb enable caused the mapping switch is still a question.   Btw, it appears Hauptwerk has a mechanism to remap the channels.

**Comment by OrgelSpieler** [comment, marked as answer]
I've determined that the channel numbering has to do with some interaction between Windows and my two RME cards. RME uses the same ASIO driver for both cards and can in fact support up to three cards working simultaneously. How Windows decides to assign channels to these cards appears to be related to a setting buried in the RME configuration. RME has a setting for the "card order." Because my AIO was second in the list, Windows assigned channels to it beginning at channel 13. However, when I had the "other" card disabled, Windows began channel assignments for the AIO card at Channel 1. It seems that Windows maintained this assignment even after the "other" card was enabled until something glitched brought on by enabling the convolution reverb. After the glitch, the channels assignments reverted to order as specified by the RME "card order." I'm not sure I'll ever be able to reproduce the problem as I now have the correct card order specified in the RME configuration, but at least I have a better understanding of how everything interacts with one another.

Even with the correct Card order and channels assignments, I still have unused channels in the Audio Output List (AES and Main) that show up in the menu bar. It would be a nice feature to be able to suppress those.

**Reply by oleg68** [reply]
> Even with the correct Card order and channels assignments, I still have unused channels in the Audio Output List (AES and Main) that show up in the menu bar. It would be a nice feature to be able to suppress those.

Could you post the screenshoot?

**Reply by OrgelSpieler** [reply]
Here are screenshots of the audio output map and menu bar.  There are a total of twenty channels defined in the audio map.  Note that channels 6-10 are not specified but have a placeholder in the map.  These are hardware interfaces  (phones and AES) on the card that I don't use.  Because the channels show in the map, 20 channels show on the menu bar rather than 16.

![MenuBar](https://user-images.githubusercontent.com/22777068/148130359-70860ef5-d954-4cad-b1f7-27c3ae73fe63.JPG)
![OuputMap](https://user-images.githubusercontent.com/22777068/148130361-be74940c-1851-47c5-aadd-08c5d463814a.JPG)

**Comment by oleg68** [comment]
@OrgelSpieler it is not a menu bar. It is a special list that allows to map audio groups to channels.

Now the only way of setting the channel number it is it's position. So deleting channels expept the last one is not possible. This list is scrollable so I cann't see any problems with unused channels.

About your first problem: if you enable the first RME card, it takes channels 1-12. I'd suggest you to change the card order, that the RME for GrandOrgue would always take channels 1-20, and the other one - 21-32.
